File: src/python/nested_dict_accessor.py
from dataclasses import dataclass
from typing import List, Any

@dataclass
class NestedDictAccess:
    data: dict
    keys: List

    # dynamically access or set nested dictionary keys in model

    @property
    def val(self) -> str:
        data = self.data
        for k in self.keys:
            data = data[k]
        return data

    @val.setter
    def val(self, v: str) -> None:
        data = self.data
        lastkey = self.keys[-1]
        for k in self.keys[:-1]:  # when assigning drill down to *second* last key
            data = data[k]
        data[lastkey] = v


class DynamicAccessNestedDict:
    """Dynamically get/set nested dictionary keys of 'data' dict"""

    def __init__(self, data: dict):
        self.data = data

    def getval(self, keys: List):
        data = self.data
        for k in keys:
            data = data[k]
        return data

    def setval(self, keys: List, val) -> None:
        data = self.data
        lastkey = keys[-1]
        for k in keys[:-1]:  # when assigning drill down to *second* last key
            data = data[k]
        data[lastkey] = val


if __name__ == "__main__":
    # Quick test of ModelRef

    model = {
        "welcomemsg": "Welcome", 
        "user": {
            "firstname": "Sam", 
            "surname": "Smith"
        }
    }

    m = NestedDictAccess(model, ["user", "firstname"])
    assert m.val == "Sam"
    m.val = "Mary"
    assert m.val == "Mary"

    model = {
        "welcomemsg": "Welcome", 
        "user": {
            "firstname": "Sam", 
            "surname": "Smith"
        }
    }

    m = DynamicAccessNestedDict(model)
    assert m.getval(["user", "firstname"]) == "Sam"
    m.setval(["user", "firstname"], "Mary")
    assert m.getval(["user", "firstname"]) == "Mary"


    # My Stack Overflow answer example - https://stackoverflow.com/questions/39818669/python-dynamically-accessing-nested-dictionary-keys

    dct = {'label': 'A', 'config': {'value': 'val1'}}

    d = DynamicAccessNestedDict(dct)
    assert d.getval(["label"]) == "A"
    assert d.getval(["config", "value"]) == "val1"

    # Set some values
    d.setval(["label"], "B")
    d.setval(["config", "value"], "val2")

    assert d.getval(["label"]) == "B"
    assert d.getval(["config", "value"]) == "val2"

    print("OK")

File: src/javascript/js/nested_dict_accessor.js
/**
 * NestedDictAccess - wraps an object 'model' so that setting .val will
 * get or set the key specified in that object. The parameter 'keys'
 * is a list of keys, which allows drilling through sub objects/dicts 
 * during that get or set access.
 * 
 * E.g. The keys ["user", "firstname"] will access model.user.firstname
 * 
 * Reason for this class? Well, it lets you record what you want to access once
 * then generic functions can later access that attribute without knowing
 * specifically what they are accessing. I suppose its like a 
 * factory function returning a lambda, where
 * you wrap some behaviour in a function and then users of that function 
 * simply call that function, not knowing the behaviour.
 */
class NestedDictAccess {  // Reference to a shared model object
    constructor(model, keys) {
      this.model = model;  // any object/dict
      this.keys = keys;  // ['a', 'b'] would refer to model.a.b
      this.finalstr = "";
    }
  
    // dynamically access or set nested dictionary keys
  
    get val() {
      let data = this.model
      for (let k of this.keys)
        data = data[k]
      return data
    }
  
    set val(val) {
      let data = this.model
      let numkeys = this.keys.length
      let lastkey = this.keys.slice(-1)
      for (let k of this.keys.slice(0, numkeys - 1))  // for assignment drill down to *second* last key
        data = data[k]
      data[lastkey] = val
    }
}
  
// Quick test of NestedDictAccess
// (uncomment all the code below this line)
// -----------------------------------------------------

function assert(condition, message) {
  if (!condition) {
      message = message || "Assertion failed";
      if (typeof Error !== "undefined") {
          throw new Error(message);
      }
      throw message; // Fallback
  }
}
var model = {
  welcomemsg: "Welcome", 
  user: {
    firstname: "Sam", 
    surname: "Smith"
  }
}
assert(model.user.firstname == "Sam")
let m_firstname = new NestedDictAccess(model, ["user", "firstname"])
assert(m_firstname.val == "Sam")
m_firstname.val = "Mary"
assert(m_firstname.val == "Mary")
assert(model.user.firstname == "Mary")

/**
 * I suppose it could be done with a functions but have to mention the 
 * object name 'obj' each time
 */
let surname_set = (obj, val) => { obj.user.surname = val }
let surname_get = (obj) => { return obj.user.surname }
// test
assert(model.user.surname == "Smith")
assert(surname_get(model) == "Smith")
surname_set(model, "Flinstone")  // change it using the lambda
assert(model.user.surname == "Flinstone")
assert(surname_get(model) == "Flinstone")

/** 
 * I suppose it could be done with lambdas, functions which return a function locked to an obj
 */ 
let factory_model_surname_set = function (obj) { return (val)=> { obj.user.surname = val }}
let factory_model_surname_get = function (obj) { return () => { return obj.user.surname }}
// create the lambda functions
let m_surname_set = factory_model_surname_set(model)
let m_surname_get = factory_model_surname_get(model)
// test
assert(model.user.surname == "Flinstone")
assert(m_surname_get() == "Flinstone")  // lambda approach, notice we don't mention 'model' its baked in
m_surname_set("Jones")  // lambda approach, again
assert(model.user.surname == "Jones")
